libraries/platforms/dromajo-vcs/bsg_manycore_simulator.cpp

```diff
@@ -0,0 +1,662 @@
+      drains++;
+  } while ((err == BSG_NONSYNTH_DPI_NOT_WINDOW || err == BSG_NONSYNTH_DPI_SUCCESS)
+          && drains <= cap);
+
+  if (drains == cap) {
+    bsg_pr_err("Failed to drain FIFO\n");
+    return HB_MC_FAIL;
+  }
+
+  return HB_MC_SUCCESS;
+}
+
+/*
+ * dpi_init
+ * Initializes the DPI between the manycore and Dromajo
+ * @params[in] dpi - Pointer to the manycore DPI object
+ * @params[in] hierarchy - A C++ string that holds the manycore hierarchy
+ * @returns success if initialized correctly
+ */
+int SimulationWrapper::dpi_init() {
+  advance_time();
+
+  int err;
+
+  err = dpi_fifo_drain(HB_MC_FIFO_RX_REQ);
+  if (err != HB_MC_SUCCESS) {
+    dpi_cleanup();
+    return err;
+  }
+
+  err = dpi_fifo_drain(HB_MC_FIFO_RX_RSP);
+  if (err != HB_MC_SUCCESS) {
+    dpi_cleanup();
+    return err;
+  }
+
+  return HB_MC_SUCCESS;
+}
+
+/*
+ * dromajo_init
+ * Initializes dromajo with the correct command-line arguments
+ * @returns a pointer to a dromajo_cosim_state_t object
+ */
+int SimulationWrapper::dromajo_init() {
+  char dromajo_str[50];
+  char host_str[50];
+  char manycore_str[50];
+  char prog_str[50];
+
+  sprintf(dromajo_str, "dromajo");
+  sprintf(host_str, "--host");
+  sprintf(manycore_str, "--manycore");
+  sprintf(prog_str, "main.elf");
+
+  char* argv[] = {dromajo_str, host_str, manycore_str, prog_str};
+  dromajo = dromajo_cosim_init(4, argv);
+  if (!dromajo) {
+    bsg_pr_err("Failed to initialize Dromajo!\n");
+    return HB_MC_FAIL;
+  }
+  return HB_MC_SUCCESS;
+}
+
+/*
+ * dromajo_step
+ * Executes 1 instruction in Dromajo
+ * @returns true if execution is incomplete and/or without errors
+ */
+bool SimulationWrapper::dromajo_step() {
+  int err = dromajo_cosim_step(dromajo, 0, 0, 0, 0, 0, false, false);
+  return (err != 0) ? false : true;
+}
+
+/*
+ * dromajo_transmit_packet
+ * Fetches data from the Dromajo->Manycore FIFO and pushes it into the DPI FIFO
+ * to send packets to the manycore
+ * @returns success on succesful transmission
+ */
+int SimulationWrapper::dromajo_transmit_packet() {
+  hb_mc_packet_t dromajo_to_mc_packet, host_to_dromajo_packet;
+  int err;
+  __m128i *pkt;
+
+  mc_fifo_type_t type = FIFO_HOST_TO_MC_REQ;
+  bool is_empty;
+
+  do {
+    // Read the FIFO head pointer for all 32-bit FIFOs
+    dromajo_to_mc_packet.words[0] = host_to_mc_req_fifo->fifo[0].front();
+    dromajo_to_mc_packet.words[1] = host_to_mc_req_fifo->fifo[1].front();
+    dromajo_to_mc_packet.words[2] = host_to_mc_req_fifo->fifo[2].front();
+    dromajo_to_mc_packet.words[3] = host_to_mc_req_fifo->fifo[3].front();
+
+    // Intercept packets that are for the host and generate appropriate responses
+    // TODO: Currently, these packets don't go over the network. In the real system, they will and the simulation infrastructure
+    // must emulate that as best as possible.
+    if ((dromajo_to_mc_packet.request.x_dst == HOST_X_COORD) && (dromajo_to_mc_packet.request.y_dst == HOST_Y_COORD)) {
```

Comment:
In the future I would factor this into two separate. methods, one for sending to the manycore and one for sending to the host. The if statement is fine, but there's a lot of code in a single method for transmitting packets.

Reply:
I kept the packet transmission to the manycore in the same function and moved the host packet handling to another function. Can break it down further if needed.